scylla-macros/src/deserialize/row.rs

```diff
@@ -0,0 +1,500 @@
+
+        let crate_path = &self.0.struct_attrs().crate_path;
+        let constraint_lifetime = self.0.constraint_lifetime();
+
+        let required_fields: Vec<_> = self.0.fields().iter().filter(|f| f.is_required()).collect();
+        let field_idents: Vec<_> = (0..required_fields.len())
+            .map(|i| quote::format_ident!("f_{}", i))
+            .collect();
+        let name_verifications: Vec<_> = required_fields
+            .iter()
+            .zip(field_idents.iter())
+            .enumerate()
+            .map(|(id, (field, fidents))| self.generate_name_verification(id, field, fidents))
+            .collect();
+
+        let field_deserializers = required_fields.iter().map(|f| f.deserialize_target());
+
+        let field_count = required_fields.len();
+
+        parse_quote! {
+            fn type_check(
+                specs: &[#crate_path::frame::response::result::ColumnSpec],
+            ) -> ::std::result::Result<(), #crate_path::frame::frame_errors::ParseError> {
+                match specs {
+                    [#(#field_idents),*] => {
+                        #(
+                            // Verify the name (unless `no_field_name_verification' is specified)
+                            #name_verifications
+
+                            // Verify the type
+                            // TODO: Provide better context about which field this error is about
+                            <#field_deserializers as #crate_path::types::deserialize::value::DeserializeCql<#constraint_lifetime>>::type_check(&#field_idents.typ)?;
+                        )*
+                        ::std::result::Result::Ok(())
+                    },
+                    _ => ::std::result::Result::Err(
+                        #crate_path::frame::frame_errors::ParseError::BadIncomingData(
+                            format!(
+                                "Wrong number of columns, expected {} but got {}",
+                                #field_count,
+                                specs.len(),
+                            )
+                        ),
+                    ),
+                }
+            }
+        }
+    }
+}
+
+struct DeserializeAssumeOrderGenerator<'sd>(&'sd StructDesc);
+
+impl<'sd> DeserializeAssumeOrderGenerator<'sd> {
+    fn generate_finalize_field(&self, field: &Field) -> syn::Expr {
+        if field.skip {
+            // Skipped fields are initialized with Default::default()
+            return parse_quote!(::std::default::Default::default());
+        }
+
+        let crate_path = self.0.struct_attrs().crate_path();
+        let deserializer = field.deserialize_target();
+        let constraint_lifetime = self.0.constraint_lifetime();
+        parse_quote!(
+            {
+                let col = row.next().ok_or_else(|| {
+                    #crate_path::frame::frame_errors::ParseError::BadIncomingData(
+                        "Not enough fields".to_owned(),
```

Comment:
This message is misleading. When `ok_or_else` fails that means that there are more fields than rows in the `ColumnIterator`.

Reply:
I'll change it to "Not enough columns", then.